<!--
  Copyright 2016 The Cartographer Authors

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->

<launch>
  <arg name="no_rviz" default="false"/>
  <arg name="bag_filenames" default=""/> 
  <param name="/use_sim_time" value="true" />
  <rosparam command="load"
      file="$(find avidbots_cartographer_mapping)/param/cartographer_mapping.yaml"/>
  <group unless="$(arg no_rviz)">
    <node name="rviz" pkg="rviz" type="rviz" required="true"
        args="-d $(find cartographer_ros)/configuration_files/demo_2d.rviz" />

    <node name="cartographer_occupancy_grid_node" pkg="cartographer_ros"
        type="cartographer_occupancy_grid_node" args="-resolution 0.05" />
  </group>

  <node name="cartographer_offline_node" pkg="cartographer_ros"
      required="$(arg no_rviz)"
      type="cartographer_offline_node" args="
          -configuration_directory $(find avidbots_cartographer_mapping)/configuration_files/cartographer_config
          -configuration_basename ../cartographer_ros_config/offline_mapping.lua
          -bag_filenames $(arg bag_filenames)"
      output="screen">
    <!-- this remapping is necessary as cartographer is expecting odometry to be published to "odom topic" -->
    <!-- <remap from="odom" to="avidbots/base/dead_reckon"/> -->

    <!-- one possible issue with using cartograher is the scan topic is a merged scan, thus
         have a large variance on publish frequency, the following two remap basically
         set the scan_front as the only scan feeding to cartograher, dump the back and merged
         scan data -->
    <!-- <remap from="whatever" to="scan" /> -->
    <remap from="scan" to="scan_front"/>
  </node>

  <node name="avidbots_cartographer_offline_mapping_node" type="avidbots_cartographer_offline_mapping_node" pkg="avidbots_cartographer_mapping" output="screen">
  </node>

</launch>
